backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/query/impl/LuceneSearchResultExtractorImpl.java

```diff
@@ -71,4 +95,58 @@
 
 		return extractedData;
 	}
+
+	private Map<Integer, Set<Integer>> fetchNestedDocs(IndexSearcher indexSearcher, ScoreDoc[] scoreDocs) throws IOException {
+		// if the projection does not need any nested object skip their fetching
+		if ( storedFieldVisitor.getNestedDocumentPaths().isEmpty() ) {
+			return new HashMap<>();
+		}
+
+		Map<String, Integer> parentIds = new HashMap<>();
+		for ( ScoreDoc hit : scoreDocs ) {
+			Document doc = indexSearcher.doc( hit.doc, ID_FIELD_SET );
```

Comment:
This could be avoided if we simply added a collector to the initial search that would collect the id using doc values.
Now that I think about it, it could be a great improvement in general: if we could collect IDs during the query execution rather than after it, we may gain some performance, especially for large indexes.
No need to address this now; I created https://hibernate.atlassian.net/browse/HSEARCH-3657

Reply:
I will add a comment over here with the Jira key.